Automattic/pocket-casts-ios
# Improve UI responsiveness of the Up Next view on the Apple Watch for large collections

Hi folks! I'm learning Swift, and my most used app happens to be open-source, so it was a no-brainer to try this.

I have a large amount of episodes on my Up Next list (~800, I just put my whole library in there), and my Apple Watch S5 struggles when doing any operation with it. Opening the Up Next view takes about 3 seconds, in which the UI is completely frozen. So I gave Instruments a try and got to hacking. I can only test it in the Simulator (my watch doesn't pair with XCode for some reason), but with all these changes, opening Up Next should be instant. If these changes are implemented for the other lists (Podcast Episodes, Filters, Files), maybe the [limit of rows](https://github.com/Automattic/pocket-casts-ios/blob/297a81580766c13005a56bdf3cba758fb5971349/podcasts/Constants.swift#L192) on the Apple Watch can be removed altogether (fixing #191).

I'll explain what every commit does, in order:
- https://github.com/Automattic/pocket-casts-ios/commit/867b24e01e354c6c5d025da8bf40b04d3c0e8b1c: For every episode in the list, [this code](https://github.com/Automattic/pocket-casts-ios/blob/297a81580766c13005a56bdf3cba758fb5971349/Pocket%20Casts%20Watch%20App%20Extension/EpisodeViewModel.swift#L15) checks if it's in UpNext. That performs a linear search on an array of `PlaylistEpisode`s. This commit adds an auxiliary data structure (a `Set`) to speed up that check. The memory impact should be minimal because it only stores the UUIDs. This commit doesn't have a big impact if the rest of the fixes are implemented, but it doesn't hurt.
- https://github.com/Automattic/pocket-casts-ios/commit/aeae11443174bdfa7517886ae45416c2702eb2e9: This is a simple optimization. We want to merge 2 arrays based on the order of a 3rd array. Because both arrays are already sorted, there's no need to do a full array scan for every insertion into the new list. The code is certainly not as elegant, but it's instant (as opposed to maybe 500ms on a real device). I'm open to rewriting it in a more "swift" way if you give me some pointers though.
- https://github.com/Automattic/pocket-casts-ios/commit/2f93b761aa4449371b60ddefd68c3cff4fe7a8fa: The big one. Most of the time when rendering the Up Next view is spend initializing all of the `EpisodeRowViewModel`s. The initialization (I called it `hydration` but I'm not married to that, naming is hard) is now postponed until the last possible moment, when the row view is going to be rendered.
- https://github.com/Automattic/pocket-casts-ios/commit/488400c17bf1f4afbce381621b32878b98162983: This extends the "hydration" concept to Episode/UserEpisode objects too. Instead of building 800+ Episode objects from database records, the Episode objects are created in a minimal way (just with the `uuid`), and they will be "filled" just in time as required. Strictly speaking, 800 calls to the database will be slower than a single big one, but those 800 calls will be spread through a large amount of time if the user decides to scroll through the whole list.

As an alternative to the "hydration" changes, I have [another branch](https://github.com/DanReyLop/pocket-casts-ios/compare/update/upnext-performance...DanReyLop:pocket-casts-ios:try/upnext-performance-lazy-rowmodel-collection) with a different approach.  It involves having a custom collection of `EpisodeRowViewModel`s that will manage fetching the full Episode objects and building the view models. The result is the same, but I feel it's a bit more clunky. I don't like that `EpisodeStub` is a whole different object than `Episode`, for example. Let me know if you see more potential in that solution and I'll give it another think.

Let me know how do you prefer to proceed. Maybe it's worth it to add some comments here and there, and I haven't even checked if tests pass, but I wanted to put this in front of y'all in case you disagree fundamentally with those changes. The first 2 commits can be merged independently, for sure, so let me know if I should open separate PRs.

## To test

- Have a big Up Next list. Adding "WTF With Marc Marron" for example should get you there, it's nearly 1000 episodes.
- On your Apple Watch, do stuff on the Up Next view: open it, pick an episode and move it up top, pick an episode and start playing it, etc.
- On `trunk`, any of those operations would result in a ~3s freeze (maybe a bit less on newer Watches). With this PR, it should be near-instant.

## Checklist
